File: src/org/elixir_lang/psi/scope/variable/MultiResolve.java
package org.elixir_lang.psi.scope.variable;

import com.intellij.openapi.util.Key;
import com.intellij.psi.*;
import com.intellij.psi.util.PsiTreeUtil;
import org.elixir_lang.psi.Arguments;
import org.elixir_lang.psi.ElixirDoBlock;
import org.elixir_lang.psi.ElixirStab;
import org.elixir_lang.psi.ElixirStabBody;
import org.elixir_lang.psi.impl.ElixirPsiImplUtil;
import org.elixir_lang.psi.operation.Match;
import org.elixir_lang.psi.scope.Variable;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static org.elixir_lang.psi.impl.ElixirPsiImplUtil.ENTRANCE;
import static org.elixir_lang.psi.impl.ProcessDeclarationsImpl.DECLARING_SCOPE;
import static org.elixir_lang.reference.Callable.IGNORED;

/**
 * Finds
 */
public class MultiResolve extends Variable {
    /*
     * CONSTANTS
     */

    private static final Key<PsiElement> LAST_BINDING_KEY = new Key<PsiElement>("LAST_BINDING_KEY");

    /*
     *
     * Static Methods
     *
     */

    /*
     * Public Static Methods
     */
    @NotNull
    private final String name;
    private final boolean incompleteCode;

    /*
     * Private Static Methods
     */
    @Nullable
    private List<ResolveResult> resolveResultList = null;

    public MultiResolve(@NotNull String name, boolean incompleteCode) {
        this.incompleteCode = incompleteCode;
        this.name = name;
    }

    @Nullable
    public static List<ResolveResult> resolveResultList(@NotNull String name,
                                                        boolean incompleteCode,
                                                        @NotNull PsiElement entrance) {
        List<ResolveResult> resolveResultList;

        if (name.equals(IGNORED)) {
            resolveResultList = Collections.<ResolveResult>singletonList(new PsiElementResolveResult(entrance));
        } else {
            resolveResultList = resolveResultList(name, incompleteCode, entrance, ResolveState.initial());
        }

        return resolveResultList;
    }

    /*
     * Fields
     */

    @Nullable
    public static List<ResolveResult> resolveResultList(@NotNull String name,
                                                        boolean incompleteCode,
                                                        @NotNull PsiElement entrance,
                                                        @NotNull ResolveState resolveState) {
        MultiResolve multiResolve = new MultiResolve(name, incompleteCode);
        ResolveState treeWalkUpResolveState = resolveState;

        if (treeWalkUpResolveState.get(ENTRANCE) == null) {
            treeWalkUpResolveState = treeWalkUpResolveState.put(ENTRANCE, entrance);
        }

        PsiTreeUtil.treeWalkUp(
                multiResolve,
                entrance,
                entrance.getContainingFile(),
                treeWalkUpResolveState
        );

        return multiResolve.getResolveResultList();
    }

    @Contract(pure = true)
    @Nullable
    private static PsiElement previousExpression(@NotNull PsiElement element) {
        PsiElement expression = ElixirPsiImplUtil.previousSiblingExpression(element);

        if (expression == null) {
            expression = previousParentExpression(element);
        }

        return expression;
    }

    @Contract(pure = true)
    @Nullable
    private static PsiElement previousParentExpression(@NotNull PsiElement element) {
        PsiElement expression = element;

        do {
            expression = expression.getParent();
        } while (expression instanceof Arguments ||
                expression instanceof ElixirDoBlock ||
                expression instanceof ElixirStab ||
                expression instanceof ElixirStabBody);

        return expression;
    }

    /*
     *
     * Instance Methods
     *
     */

    /*
     * Public Instance Methods
     */

    @Nullable
    public List<ResolveResult> getResolveResultList() {
        return resolveResultList;
    }

    /*
     * Protected Instance Methods
     */

    /**
     * Decides whether {@code match} matches the criteria being searched for.  All other {@link #executeOnVariable} methods
     * eventually end here.
     *
     * @param match
     * @param state
     */
    @Override
    protected boolean executeOnVariable(@NotNull PsiNamedElement match, @NotNull ResolveState state) {
        addToResolveResultListIfMatchingName(match, state);

        return org.elixir_lang.psi.scope.MultiResolve.keepProcessing(incompleteCode, resolveResultList);
    }

    /*
     * Private Instance Methods
     */

    /**
     * Adds {@code resolveResult} to {@link #resolveResultList} if it exists; otherwise, create new
     * {@code List<ResolveList>}, set {@link #resolveResultList} to it, and add {@code resolveResult} to it.
     *
     * @param resolveResult The element to add to {@code resolveResultList}
     */
    private void addToResolveResultList(@NotNull ResolveResult resolveResult) {
        if (resolveResultList == null) {
            resolveResultList = new ArrayList<ResolveResult>();
        }

        resolveResultList.add(resolveResult);
    }

    private void addToResolveResultList(@NotNull PsiElement element, ResolveState state, boolean validResult) {
        Boolean declaringScope = state.get(DECLARING_SCOPE);

        if (declaringScope == null || declaringScope) {
            PsiElement lastBinding = state.get(LAST_BINDING_KEY);
            boolean added = false;

            /* if LAST_BINDING_KEY is set, then we're checking if a right-hand match is bound higher up, so an effective
               recursive call.  If the recursive call got the same result, stop the recursion by not checking for
               rebinding */
            if (lastBinding == null || !element.isEquivalentTo(lastBinding)) {
                Match matchAncestor = PsiTreeUtil.getContextOfType(element, Match.class);

                if (matchAncestor != null) {
                    PsiElement rightOperand = matchAncestor.rightOperand();

                    if (rightOperand != null) {
                        /* right-hand match can only be declarative if it is not already bound, so need to try to
                           resolve further up to try to find if {@code element} is already bound */
                        if (PsiTreeUtil.isAncestor(rightOperand, element, false)) {
                            // previous sibling or parent to search for earlier binding
                            PsiElement expression = previousExpression(matchAncestor);

                            if (expression != null) {
                                List<ResolveResult> preboundResolveResultList = resolveResultList(
                                        name,
                                        incompleteCode,
                                        expression,
                                        ResolveState
                                                .initial()
                                                .put(ENTRANCE, matchAncestor)
                                                .put(LAST_BINDING_KEY, element)
                                );

                                if (preboundResolveResultList != null && preboundResolveResultList.size() > 0) {
                                    if (resolveResultList == null) {
                                        resolveResultList = preboundResolveResultList;
                                    } else {
                                        resolveResultList.addAll(preboundResolveResultList);
                                    }

                                    added = true;
                                }
                            }
                        }
                    }
                }
            }

            // either non-right match declaration or recursive call didn't find a rebinding
            if (!added) {
                addToResolveResultList(new PsiElementResolveResult(element, validResult));
            }
        }
    }

    private void addToResolveResultListIfMatchingName(@NotNull PsiNamedElement match, ResolveState state) {
        String matchName = match.getName();

        if (matchName != null) {
            if (matchName.equals(name)) {
                addToResolveResultList(match, state, true);
            } else if (incompleteCode && matchName.startsWith(name)) {
                addToResolveResultList(match, state, false);
            }
        }
    }
}

File: src/org/elixir_lang/psi/scope/variable/MultiResolve.kt
package org.elixir_lang.psi.scope.variable

import com.intellij.openapi.project.DumbService
import com.intellij.openapi.util.Key
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiNamedElement
import com.intellij.psi.ResolveResult
import com.intellij.psi.ResolveState
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.stubs.StubIndex
import com.intellij.psi.util.PsiTreeUtil
import org.elixir_lang.psi.*
import org.elixir_lang.psi.call.Call
import org.elixir_lang.psi.ex_unit.Assertions
import org.elixir_lang.psi.impl.ElixirPsiImplUtil.ENTRANCE
import org.elixir_lang.psi.impl.ElixirPsiImplUtil.previousSiblingExpression
import org.elixir_lang.psi.impl.ProcessDeclarationsImpl.DECLARING_SCOPE
import org.elixir_lang.psi.operation.Match
import org.elixir_lang.psi.scope.MultiResolve.keepProcessing
import org.elixir_lang.psi.scope.Variable
import org.elixir_lang.psi.scope.VisitedElementSetResolveResult
import org.elixir_lang.psi.stub.index.QuoteVariableName
import org.elixir_lang.reference.Callable
import org.jetbrains.annotations.Contract

class MultiResolve(private val name: String, private val incompleteCode: Boolean) : Variable() {
    private val resolveResultList = mutableListOf<VisitedElementSetResolveResult>()

    /**
     * Decides whether `match` matches the criteria being searched for.  All other [.executeOnVariable] methods
     * eventually end here.
     *
     * @param match
     * @param state
     */
    override fun executeOnVariable(match: PsiNamedElement, state: ResolveState): Boolean {
        addToResolveResultListIfMatchingName(match, state)
        return keepProcessing(incompleteCode, resolveResultList)
    }

    private fun addToResolveResultList(element: PsiElement, state: ResolveState, validResult: Boolean) {
        if (state.get(DECLARING_SCOPE) == false) return

        /* A read on the right of `=` declares its name only if nothing above binds it, so its earlier bindings are
           looked up: not when this is that lookup meeting the read again, and not for a prefix-named read, which is
           a candidate for incomplete code and never a declaration of the searched name. */
        val lastBinding = state.get(LAST_BINDING_KEY)
        val earlier = if (validResult && (lastBinding == null || !element.isEquivalentTo(lastBinding))) {
            earlierBindingResults(name, incompleteCode, element, state)
        } else {
            emptyList()
        }
        val bound = earlier.filter { it.isValidResult }

        if (bound.isNotEmpty()) {
            resolveResultList.addAll(bound)
        } else {
            resolveResultList.add(VisitedElementSetResolveResult(element, validResult, state.visitedElementSet()))
        }
        // a prefix-named earlier variable is a candidate for incomplete code, whether or not the read is bound
        if (incompleteCode) {
            resolveResultList.addAll(earlier.filter { !it.isValidResult })
        }
    }

    /** The results of resolving [name] just before the match [element] sits on the right of; empty elsewhere. */
    private fun earlierBindingResults(
        name: String,
        incompleteCode: Boolean,
        element: PsiElement,
        state: ResolveState
    ): List<VisitedElementSetResolveResult> {
        val match = PsiTreeUtil.getContextOfType(element, Match::class.java) ?: return emptyList()
        val rightOperand = match.rightOperand() ?: return emptyList()
        // a pattern on the right, an `fn` parameter or a clause head, binds afresh rather than reading
        val pattern = BindingPattern.of(element)

        return if (PsiTreeUtil.isAncestor(rightOperand, element, false) &&
            (pattern == null || !PsiTreeUtil.isAncestor(rightOperand, pattern, false))
        ) {
            resolveBefore(name, incompleteCode, match, element, state)
        } else {
            emptyList()
        }
    }

    private fun addToResolveResultListIfMatchingName(match: PsiNamedElement, state: ResolveState) {
        match.name?.let { name ->
            if (name.startsWith(this.name)) {
                val validResult = name == this.name
                addToResolveResultList(match, state, validResult)
            }
        }
    }

    companion object {
        private val LAST_BINDING_KEY = Key<PsiElement>("LAST_BINDING_KEY")

        /** Resolves [name] from just before [match], as a read there would, on behalf of [element] inside it. */
        private fun resolveBefore(
            name: String,
            incompleteCode: Boolean,
            match: Match,
            element: PsiElement,
            state: ResolveState
        ): List<VisitedElementSetResolveResult> {
            val expression = previousExpression(match, state) ?: return emptyList()
            val searchState = ResolveState.initial()
                .put(ENTRANCE, match)
                .putInitialVisitedElement(match)
                .putVisitedElements(state)
                .put(LAST_BINDING_KEY, element)

            return resolveInScope(name, incompleteCode, expression, searchState)
        }

        fun resolveResultList(name: String,
                              incompleteCode: Boolean,
                              entrance: PsiElement): List<VisitedElementSetResolveResult> =
            if (name == Callable.IGNORED) {
                listOf<VisitedElementSetResolveResult>(VisitedElementSetResolveResult(entrance, true, emptySet()))
            } else {
                val resolveState = ResolveState.initial().put(ENTRANCE, entrance).putInitialVisitedElement(entrance)

                resolveInScope(name, incompleteCode, entrance, resolveState)
                        .takeIf { set -> set.any(ResolveResult::isValidResult) }
                        ?: nameInAnyQuote(entrance, name, incompleteCode)
            }

        /**
         * The declarations of [name] that a read placed just before [declaration]'s match resolves to: what an
         * assignment there rebinds. Empty when [declaration] is not bound by a match.
         */
        fun earlierBindings(name: String, declaration: PsiElement): List<PsiElement> {
            val match = PsiTreeUtil.getContextOfType(declaration, Match::class.java) ?: return emptyList()
            val state = ResolveState.initial().put(ENTRANCE, match).putInitialVisitedElement(match)

            return resolveBefore(name, false, match, declaration, state)
                .mapNotNull { result -> result.element.takeIf { result.isValidResult } }
        }

        fun resolveInScope(name: String,
                           incompleteCode: Boolean,
                           entrance: PsiElement,
                           resolveState: ResolveState): List<VisitedElementSetResolveResult> {
            val multiResolve = MultiResolve(name, incompleteCode)

            val treeWalkUpResolveState = if (resolveState.get(ENTRANCE) == null) {
                resolveState.put(ENTRANCE, entrance)
            } else {
                resolveState
            }

            PsiTreeUtil.treeWalkUp(
                    multiResolve,
                    entrance,
                    entrance.containingFile,
                    treeWalkUpResolveState
            )

            return multiResolve.resolveResultList
        }

        private fun nameInAnyQuote(entrance: PsiElement,
                                   name: String,
                                   incompleteCode: Boolean): List<VisitedElementSetResolveResult> {
            val project = entrance.project
            val resolveResults = mutableListOf<VisitedElementSetResolveResult>()

            if (!DumbService.isDumb(project)) {
                val stubIndex = StubIndex.getInstance()
                val keys = mutableListOf<String>()

                stubIndex.processAllKeys(QuoteVariableName.KEY, project) { key ->
                    if ((incompleteCode && key.startsWith(name)) || key == name) {
                        keys.add(key)
                    }

                    true
                }

                val scope = GlobalSearchScope.allScope(project)
                // results are never valid because the qualifier is unknown
                val validResult = false

                for (key in keys) {
                    stubIndex
                            .processElements(QuoteVariableName.KEY, key, project, scope, NamedElement::class.java) { namedElement ->
                                resolveResults.add(VisitedElementSetResolveResult(namedElement, validResult, emptySet()))

                                true
                            }
                }
            }

            return resolveResults
        }

        @Contract(pure = true)
        private fun previousExpression(element: PsiElement, state: ResolveState): PsiElement? =
            previousSiblingExpression(element) ?: previousParentExpression(element, state)

        @Contract(pure = true)
        private fun previousParentExpression(element: PsiElement, state: ResolveState): PsiElement? {
            var expression = element
            do {
                expression = expression.parent
            } while (expression is Arguments ||
                    expression is ElixirDoBlock ||
                    expression is ElixirStab ||
                    expression is ElixirStabBody || (expression is Call && Assertions.isChild(expression, state)))
            return expression
        }
    }
}
